scenario-compiler/src/main/scala/pl/touk/nussknacker/engine/extension/Conversion.scala

```diff
@@ -0,0 +1,85 @@
+package pl.touk.nussknacker.engine.extension
+
+import cats.data.ValidatedNel
+import cats.implicits.catsSyntaxValidatedId
+import pl.touk.nussknacker.engine.api.generics.GenericFunctionTypingError
+import pl.touk.nussknacker.engine.api.typed.typing.{Typed, TypingResult}
+import pl.touk.nussknacker.engine.util.classes.Extensions.ClassExtensions
+
+import java.lang.{Boolean => JBoolean, Number => JNumber}
+import java.math.{BigDecimal => JBigDecimal, BigInteger => JBigInteger}
+import java.util.{Collection => JCollection}
+import scala.util.{Success, Try}
+
+trait Conversion {
+  type ResultType >: Null <: AnyRef
+  val resultTypeClass: Class[ResultType]
+
+  def convertEither(value: Any): Either[Throwable, ResultType]
+  def appliesToConversion(clazz: Class[_]): Boolean
+
+  def canConvert(value: Any): JBoolean = convertEither(value).isRight
+
+  def convert(value: Any): ResultType = convertEither(value) match {
+    case Right(value) => value
+    case Left(ex)     => throw ex
+  }
+
+  def convertOrNull(value: Any): ResultType = convertEither(value) match {
+    case Right(value) => value
+    case Left(_)      => null
+  }
+
+  def typingResult: TypingResult = Typed.typedClass(resultTypeClass)
+  def typingFunction(invocationTarget: TypingResult): ValidatedNel[GenericFunctionTypingError, TypingResult] =
+    typingResult.validNel
+}
+
+trait NumericConversion extends Conversion {
```

Comment:
Let's rename it to `ToNumericConversion`. I think that it is the last one

Reply:
Done